openedx/core/djangoapps/user_api/management/commands/migrate_user_profile_langs.py

```diff
@@ -0,0 +1,102 @@
+from openedx.core.djangoapps.user_api.models import UserPreference
+
+
+DEFAULT_CHUNK_SIZE = 10000
+DEFAULT_SLEEP_TIME_SECS = 10
+
+LOGGER = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    """
+    Implementation of the migrate command
+    """
+    help = 'Migrate all user language preferences (excluding dark languages) from one language code to another.'
+
+    def add_arguments(self, parser):
+        parser.add_argument('old_lang_code',
+                            help='Original language code, ex. "zh-cn"')
+        parser.add_argument('new_lang_code',
+                            help='New language code, ex. "zh-hans"')
+        parser.add_argument('--start_id',
+                            type=int,
+                            default=1,
+                            help='ID to begin from, in case a run needs to be restarted from the middle.')
+        parser.add_argument('--chunk_size',
+                            type=int,
+                            default=DEFAULT_CHUNK_SIZE,
+                            help='Number of users whose preferences will be updated per batch.')
+        parser.add_argument('--sleep_time_secs',
+                            type=int,
+                            default=DEFAULT_SLEEP_TIME_SECS,
+                            help='Number of users whose preferences will be updated per batch.')
+
+    def handle(self, *args, **options):
+        """
+        Execute the command.
+        """
+        old_lang_code = options['old_lang_code']
+        new_lang_code = options['new_lang_code']
+        chunk_size = options['chunk_size']
+        sleep_time_secs = options['sleep_time_secs']
+        start = options['start_id']
+        end = start + chunk_size
+
+        # Make sure we're changing to a code that actually exists. Presumably it's safe to move away from a code that
+        # doesn't.
+        langs = [lang_code[0] for lang_code in settings.LANGUAGES]
+        langs += DarkLangConfig.current().released_languages_list
+
+        if new_lang_code not in langs:
+            raise CommandError('{} is not a configured language code in settings.LANGUAGES '
+                               'or the current DarkLangCongig.'.format(new_lang_code))
+
+        max_id = UserPreference.objects.all().aggregate(Max('id'))['id__max']
+
+        print('Updating user language preferences from {} to {}. '
+              'Start id is {}, current max id is {}. '
+              'Chunk size is of {}'.format(old_lang_code, new_lang_code, start, max_id, chunk_size))
+
+        updated_count = 0
+
+        while True:
+            # On the last time through catch any new rows added since this run began
+            if end >= max_id:
+                print('Last round, includes all new rows added since this run started.')
+                id_query = Q(id__gte=start)
+            else:
+                id_query = Q(id__gte=start) & Q(id__lt=end)
+
+            with transaction.atomic():
```

Comment:
The following update should be performed as a single SQL query; do we need the additional transaction wrapper?

Reply:
I believe that management commands will open a transaction at db connection time, so without this it will try to execute the whole command in a single long-running transaction.

Reply:
Are you sure you aren't thinking of migrations instead?  As far as I know, there's no automatic transaction inserted around management command `handle()` methods...and even if there was, I'd expect this to just create a savepoint within it rather than break out of the transaction.

Reply:
Nope, I was thinking of previous problems I've had with long running commands, but those were in an environment where autocommit was off... and yeah this wouldn't have worked there anyway. This is currently a no-op, but I'll remove to prevent spreading my confusion around.